### PR TITLE
#98 - implemented encoding, estimate and deployment for contract creation and call

### PR DESCRIPTION
**Pull request by biboran**
To implement #98 I separated transaction bytes computation into three separate purposes (contract deployment, contract function call, direct ether transaction). I added separate encoding for contract as `EncodedContract` and function as `EncodedABIFunction`. I added computation for contract address as `ComputedContractAddress` since ganache doesn't report contract address in the `eth_getTransactionByHash` call. 

To solve #100 I added a missing check in `SimpleRLP`.

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#101](https://github.com/blockstoreapp/web3swift/issues/101) is now in scope, role is `REV`

File: Example/Tests/RLP/RLPTests.swift [hunk 1]
```diff
@@ -105,6 +105,98 @@ final class RLPTests: XCTestCase {
         )
     }
 
+    func test127() {
```
Comment (driver733) [inline Example/Tests/RLP/RLPTests.swift]: @biboran Are double `expect`s necessary?

File: Example/Tests/TransactionBytes/EthContractCallBytesTests.swift [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+import CryptoSwift
+import Nimble
+import Quick
+@testable import Web3Swift
+
+final class EthContractCallBytesTests: XCTestCase {
+
+    func testContractCallIsEncodedCorrectly() {
+        expect{
+            try EthContractCallBytes(
```
Comment (driver733) [inline Example/Tests/TransactionBytes/EthContractCallBytesTests.swift]: @biboran Seven ctor parameters seems to be too much. What do you think about splitting the tested class into smaller ones?

File: Example/Tests/TransactionBytes/EthContractCreationBytesTests.swift [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+limitations under the License.
+*/
+
+import CryptoSwift
+import Nimble
+import Quick
+@testable import Web3Swift
+
+final class EthContractCreationBytesTests: XCTestCase {
+
```
Comment (driver733) [inline Example/Tests/TransactionBytes/EthContractCreationBytesTests.swift]: @biboran Seven ctor parameters seems to be too much. What do you think about splitting the tested class into smaller ones?

File: Web3Swift/TransactionBytes/EthContractCallBytes.swift [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import Foundation
+
+//Bytes of a signed contract function call transaction
+public final class EthContractCallBytes: BytesScalar {
```
Comment (driver733) [inline Web3Swift/TransactionBytes/EthContractCallBytes.swift]: @biboran Eight ctor parameters seems to be too much. What do you think about splitting the tested class into smaller ones?

File: Web3Swift/TransactionBytes/EthContractCreationBytes.swift [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import Foundation
+
+//Bytes of a signed contract deployment transaction
+public final class EthContractCreationBytes: BytesScalar {
```
Comment (driver733) [inline Web3Swift/TransactionBytes/EthContractCreationBytes.swift]: @biboran Seven ctor parameters seems to be too much. What do you think about splitting the class into smaller ones?

File: Web3Swift/TransactionBytes/EthDirectTransactionBytes.swift [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import Foundation
+
+//Bytes of a signed transaction to ethereum address
+public final class EthDirectTransactionBytes: BytesScalar {
```
Comment (driver733) [inline Web3Swift/TransactionBytes/EthDirectTransactionBytes.swift]: @biboran Seven ctor parameters seems to be too much. What do you think about splitting the class into smaller ones?

File: Example/Tests/TransactionBytes/EthContractCallBytesTests.swift (hunk 2 above)
Comment (biboran) [reply Example/Tests/TransactionBytes/EthContractCallBytesTests.swift]: @driver733 part of the requirement for the transaction signing is an ability to reproduce it by specifying every dependency. There are 7-8 mathematically separate dependencies for the transaction signing.

Do you see any way to retain that functionality and reduce the number of parameters?

File: Web3Swift/RemoteProcedure/VerifiedProcedure.swift [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+    var description: String {
+        return error["error"]["message"].string ?? "There was no error in JSON. Here is the dump instead: \(error.debugDescription)"
+    }
+
+}
+
+//FIXME: This is a temporary workaround to allow throwing errors communicated by network. It is definitely possible to make this check perform only when the "result" was not found by reimplementing JSON.
+//Procedure that throws in case JSON RPC responded with an error
+public final class VerifiedProcedure: RemoteProcedure {
```
[the remark this answers is not in the document]
Comment (biboran) [reply Web3Swift/RemoteProcedure/VerifiedProcedure.swift]: @rockfridrich this is a temporary workaround as you can see is mentioned in the above fixme. I described the issue here: https://github.com/BlockStoreApp/Web3Swift/issues/58

**Comment by rockfridrich** [comment]
@biboran there are a lot of objects like `SimpleBytes`, `BigEndianNumber`, `CachedNumber` and it is not clear how to use them. 
We will need to have some wrappers later because it is madness to construct such objects
screen shot 2018-04-24 at 16 02 48

File: Web3Swift/TransactionBytes/EthContractCallBytes.swift [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+        - functionCall: encoded function call
+    */
+    public convenience init(
+        network: Network,
+        senderKey: PrivateKey,
+        contractAddress: BytesScalar,
+        weiAmount: NumberScalar,
+        functionCall: BytesScalar
+    ) {
+        let senderAddress = CachedBytes(
```
[the remark this answers is not in the document]
Comment (biboran) [reply Web3Swift/TransactionBytes/EthContractCallBytes.swift]: @rockfridrich makes sure `senderAddress` is computed only once. Take a look at CachedBytes [implementation.](https://github.com/BlockStoreApp/Web3Swift/blob/develop/Web3Swift/BytesScalar/CachedBytes.swift)

**Comment by biboran** [comment]
@rockfridrich the point of this ctor is to handle the complexity of signing a transaction. You only need to pass 5 vital parameters and everything else is handled inside. It might be a better idea to take it out and put into another object but I am not sure what it should be and I'll need help with that design. 

If you can start with #96 and #99 you will gain an understanding of the design and you can tell me how it can be improved and where it's extremely vague.

**Comment by 0crat** [comment]
This pull request [#101](https://github.com/blockstoreapp/web3swift/issues/101) is assigned to @driver733[/z](https://www.0crat.com/u/driver733), here is [why](https://www.0crat.com/footprint/G930TRU1E/1141358156); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @biboran[/z](https://www.0crat.com/u/biboran) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewerWe should be aware that driver733 is on vacation; this ticket may be delayed

**Comment by 0crat** [comment]
Manual assignment of issues is discouraged, see [§19](http://www.zerocracy.com/policy.html#19): -5 point(s) just awarded to @biboran[/z](https://www.0crat.com/u/biboran)

**Comment by 0crat** [comment]
@driver733[/z](https://www.0crat.com/u/driver733) this job was assigned to you 5days ago. It will be taken away from you soon, unless you close it, see [§8](http://www.zerocracy.com/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Comment by driver733** [comment]
@0crat waiting

**Comment by 0crat** [comment]
> @0crat waiting ([here](https://github.com/BlockStoreApp/Web3Swift/issues/101#issuecomment-385437238))

@driver733 It's a code review job [#101](https://github.com/blockstoreapp/web3swift/issues/101), can't put it on hold

File: Example/Tests/TransactionBytes/EthContractCallBytesTests.swift (hunk 2 above)
Comment (rockfridrich) [reply Example/Tests/TransactionBytes/EthContractCallBytesTests.swift]: @biboran @driver733 lets accept that some functions require that number of parameters due to Ethereum architecture. We could code domain wrappers for them later after experimenting with real DApps. If it is ok then let's merge this.

**Comment by 0crat** [comment]
The user @driver733[/z](https://www.0crat.com/u/driver733) resigned from [#101](https://github.com/blockstoreapp/web3swift/issues/101), please stop working. Reason for job resignation: It is older than 10 days, see [§8](http://www.zerocracy.com/policy.html#8)

**Comment by 0crat** [comment]
Resigned on delay, see [§8](http://www.zerocracy.com/policy.html#8): -15 point(s) just awarded to @driver733[/z](https://www.0crat.com/u/driver733)

**Comment by 0crat** [comment]
@biboran[/z](https://www.0crat.com/u/biboran) everybody who has role `REV` are banned at this job; I won't be able to assign anyone automatically; consider assigning someone manually (as in [§19](http://www.zerocracy.com/policy.html#19)), or invite more people (as in [§51](http://www.zerocracy.com/policy.html#51)), or remove the job from the scope (as in [§14](http://www.zerocracy.com/policy.html#14))

**Comment by 0crat** [comment]
This pull request [#101](https://github.com/blockstoreapp/web3swift/issues/101) is assigned to @driver733[/z](https://www.0crat.com/u/driver733), here is [why](https://www.0crat.com/footprint/G930TRU1E/1271445443); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @biboran[/z](https://www.0crat.com/u/biboran) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewerWe should be aware that driver733 is on vacation; this ticket may be delayed

**Comment by 0crat** [comment]
Manual assignment of issues is discouraged, see [§19](http://www.zerocracy.com/policy.html#19): -5 point(s) just awarded to @biboran[/z](https://www.0crat.com/u/biboran)

**Comment by 0crat** [comment]
Order was finished: +15 point(s) just awarded to @driver733[/z](https://www.0crat.com/u/driver733)

**Comment by 0crat** [comment]
The job [#101](https://github.com/blockstoreapp/web3swift/issues/101) is now out of scope

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @biboran[/z](https://www.0crat.com/u/biboran)

**Pull request merged**